msg_helpers.go

```diff
@@ -681,3 +681,129 @@ func packDataDomainNames(names []string, msg []byte, off int, compression compre
 	}
 	return off, nil
 }
+
+func packDataApl(data []APLPrefix, msg []byte, off int) (int, error) {
+	var err error
+	for i := range data {
+		off, err = packDataAplPrefix(&data[i], msg, off)
+		if err != nil {
+			return len(msg), err
+		}
+	}
+	return off, nil
+}
+
+func packDataAplPrefix(p *APLPrefix, msg []byte, off int) (int, error) {
+	if len(p.Network.IP) != len(p.Network.Mask) {
+		return len(msg), &Error{err: "address and mask lengths don't match"}
+	}
+
+	var family uint16
+	switch len(p.Network.IP) {
+	case net.IPv4len:
+		family = 1
+	case net.IPv6len:
+		family = 2
+	default:
+		return len(msg), &Error{err: "unrecognized address family"}
+	}
+	prefix, _ := p.Network.Mask.Size()
+	addr := p.Network.IP.Mask(p.Network.Mask)[:(prefix+7)/8]
+
+	// Write ADDRESSFAMILY
+	off, err := packUint16(family, msg, off)
+	if err != nil {
+		return len(msg), err
+	}
+	// Write PREFIX
+	off, err = packUint8(uint8(prefix), msg, off)
+	if err != nil {
+		return len(msg), err
+	}
+	// Write N and AFDLENGTH
+	var n uint8
+	if p.Negation {
+		n = 0x80
+	}
+	adflen := uint8(len(addr)) & 0x7f
+	off, err = packUint8(n|adflen, msg, off)
+	if err != nil {
+		return len(msg), err
+	}
+	// Write AFDPART
+	if off+len(addr) > len(msg) {
+		return len(msg), &Error{err: "overflow packing APL prefix"}
+	}
+	off += copy(msg[off:], addr)
+
+	return off, nil
+}
+
+func unpackDataApl(msg []byte, off int) ([]APLPrefix, int, error) {
+	var result []APLPrefix
+	for off < len(msg) {
+		prefix, end, err := unpackDataAplPrefix(msg, off)
+		if err != nil {
+			return nil, len(msg), err
+		}
+		off = end
+		result = append(result, prefix)
+	}
+	return result, off, nil
+}
+
+func unpackDataAplPrefix(msg []byte, off int) (APLPrefix, int, error) {
+	// Read ADDRESSFAMILY
+	family, off, err := unpackUint16(msg, off)
+	if err != nil {
+		return APLPrefix{}, len(msg), &Error{err: "overflow unpacking APL prefix"}
+	}
+	var ipLen int
+	switch family {
+	case 1:
+		ipLen = net.IPv4len
+	case 2:
+		ipLen = net.IPv6len
+	default:
+		return APLPrefix{}, len(msg), &Error{err: "unrecognized APL address family"}
+	}
+	// Read PREFIX
+	prefix, off, err := unpackUint8(msg, off)
```

Comment:
read prefix before the `switch family` and then do the `int(prefix)> *ipLen` check there, this way to you drop the `ipLen` variable.

I see `ipLen` is also used as the very last thing in this function, probably with building up a APLPrefix as you go, to not have to variables, but just initialize an APLPrefix along the way.

Reply:
I'm not sure I understood exactly what you proposed. I moved all the reads of the fixed-size header to the beginning of the function so that the address length handling logic is together. `ipLen` is gone and so is `neg`. Please, take a look.